### PR TITLE
Remote DebugGUI

### PR DESCRIPTION
**Pull request by luigifusco**
Add a function to render a frame as a valid JSON encoded string containing draw data.
Add a function writing to a stream a group of frames as a video.

### CODE REVIEW

File: DebugGUI/src/DebugGUI.cxx [hunk 1]
```diff
@@ -51,9 +54,80 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
...
+    }
+    first = false;
+
+    stream << getFrameJSON(draw_data);
+
+    if (curr_frame/framerate < max_frames) {
+      stream << ",\n";
+    } else {
+      stream << "]" << std::endl;
+      std::cout << "Video recorded!" << std::endl;
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:122]:
```suggestion
```

File: DebugGUI/src/DebugGUI.cxx [hunk 2]
```diff
@@ -9,6 +9,9 @@
 #include <GLFW/glfw3.h>
 #include <cstdio>
 #include <functional>
+#include <iostream>
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:12]:
```suggestion
```

File: DebugGUI/src/DebugGUI.cxx [hunk 3]
```diff
@@ -51,9 +54,80 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
   io.Fonts->AddFontFromMemoryTTF((void*)s_iconsFontAwesomeTtf, sizeof(s_iconsFontAwesomeTtf), 12.0f, &icons_config, icons_ranges);
 
   ImPlot::CreateContext();
+
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:57]:
```suggestion
```

File: DebugGUI/src/DebugGUI.cxx [hunk 4]
```diff
@@ -51,9 +54,80 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
...
   return window;
 }
 
+// returns a string containing drawing data in JSON format
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:61]:
```suggestion
/// @return a string containing drawing data in JSON format
/// FIXME: document the actual schema of the format.
```

File: DebugGUI/src/DebugGUI.cxx [hunk 5]
```diff
@@ -51,9 +54,80 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
...
+// the returned JSON is composed of a list of draw commands
+std::string getFrameJSON(ImDrawData *draw_data) {
+  std::ostringstream json_data;
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:64]: This means you will create the json_data at each frame. I would pass an ostream& as an argument.

File: DebugGUI/src/DebugGUI.cxx [hunk 6]
```diff
@@ -51,9 +54,80 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
...
+
+  json_data << "[";
+
+  for (int cmd_id = 0; cmd_id < draw_data->CmdListsCount; ++cmd_id) {
+    const auto cmd_list = draw_data->CmdLists[cmd_id];
+    const auto vtx_buffer = cmd_list->VtxBuffer;
+    const auto idx_buffer = cmd_list->IdxBuffer;
+    const auto cmd_buffer = cmd_list->CmdBuffer;
+
+
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:74]:
```suggestion
```

**Comment by ktf** [comment]
Have a look at the comments and either commit the suggestion or explain why you did it that way. We can then merge the PR after tests pass.

**Comment by luigifusco** [comment]
Separate pollGUI functionality into separate functions. Note that atm pollGUI_gl_init needs to be called every time before pollGUI_render, as it setups the window size and other stuff through ImGui_ImplGlfwGL3_NewFrame.
Add remote folder containing the png texture, the latest version of three.js and the remote access webapp.

File: DebugGUI/src/DebugGUI.cxx [hunk 7]
```diff
@@ -54,31 +55,98 @@ void* initGUI(const char* name, void(*error_callback)(int, char const*descriptio
...
+      if (i < cmd_buffer.size() - 1) json_data << ",";
+    }
+    json_data << "]}";
+    if (cmd_id < draw_data->CmdListsCount - 1) json_data << ",";
+  }
+
+  json_data << "]";
+}
+
+bool pollGUI_gl_init(void* context)
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.cxx:100]: Notice that as a convention in general we use `camelCase` unless the code comes from a third party, so here it should be:

```suggestion
bool pollGUIPreRender(void* context)
```

Also avoid referring to the graphics API as you will need to replicate this also in the Metal backend.

File: DebugGUI/src/DebugGUI.h [hunk 8]
```diff
@@ -19,6 +20,10 @@ void default_error_callback(int, const char*);
 
 void* initGUI(const char* name, decltype(default_error_callback) = nullptr);
 bool pollGUI(void* context, std::function<void(void)> guiCallback);
+void getFrameJSON(void *data, std::ostream& json_data);
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.h:23]: This you should be able to make static.

File: remote/remote.html [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+                                geometry.setDrawRange(curr_base, cmd.cnt);
+                                renderer.render( scene, camera );
+                                curr_base += cmd.cnt;
+                            }
+                            picture_counter = (picture_counter+1) % data.length;
+                        }
+                        frame_counter++;
+                    }
+
+                    animate();*/
```
Comment (ktf) [inline remote/remote.html:176]: As a general remark, do not comment code. If you  are not planning to use it immediately, simply move it into a separate function. If you do not plan to use it, drop it.

File: DebugGUI/src/DebugGUI.h (hunk 8 above)
Comment (luigifusco) [reply DebugGUI/src/DebugGUI.h:23]: My idea was to call it in O2 to get the json data from the draw data and then send it to the remote GUIs.

File: DebugGUI/src/DebugGUI.h [hunk 10]
```diff
@@ -11,6 +11,7 @@
 #define FRAMEWORK_DEBUGGUI_H
 
 #include <functional>
+#include <ostream>
```
Comment (ktf) [inline DebugGUI/src/DebugGUI.h:14]:
```suggestion
#include <iosfwd>
```

File: DebugGUI/src/DebugGUI.h (hunk 8 above)
Comment (ktf) [reply DebugGUI/src/DebugGUI.h:23]: nevermind, I missed I do not have a wrapper class for that.

**Review by ktf — CHANGES_REQUESTED** [review]
A few more changes and then we should be there.

File: DebugGUI/src/DebugGUI.h (hunk 10 above)
Comment (luigifusco) [reply DebugGUI/src/DebugGUI.h:14]: I need ostream for the stream operator

Comment (ktf) [reply DebugGUI/src/DebugGUI.h:14]: Don't you just need the implementation in the .h?

**Comment by ktf** [comment]
Ok, you probably need to implement the same API also on the macOS version (and maybe in the dummy one).

**Pull request merged**